codegen/client.go

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package codegen
+
+import (
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// ClientConfigBase is the struct for client-config.yaml. It is the base for
+// specific type of client.
+type ClientConfigBase struct {
+	Name         string       `yaml:"name" json:"name"`
+	Type         string       `yaml:"type" json:"type"`
+	Dependencies Dependencies `yaml:"dependencies" json:"dependencies"`
```

Comment:
It would probably be a lot cleaner if you just defined a `UnmarshalYAML` method that did the switch statement rather than these rather unwieldy nested structs. Then you can either provide three pointers on this struct `*HTTPClientConfig`, `*TChannelClientConfig`, `*CustomClientConfig` or you can provide the interface type you defined, whichever makes the implementation easier.